### PR TITLE
Remove Linux only Docker build for E2E tests

### PR DESCRIPTION
**Pull request by ronaldbarendse**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

### Description
While working on adding acceptance (E2E) tests to our commercial products, I noticed a big difference in build steps between Linux (using Docker and SQLite) and Windows (local project and SQL Server LocalDB). I managed to align most of the build steps (removing the Docker build on Linux), so it's easier to maintain, _faster_ and now also tests both SQL Server and SQLite databases on Windows.

The biggest issue to tackle was ensuring the Umbraco application on Linux wasn't terminating/hanging the build step, since starting a new process on Linux will by default create a child-process and Azure Pipelines waits for the main process to exit before continuing. This is fixed by using the `nohup` (no hang up) command, as mentioned in the [Start-Process documentation](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.management/start-process?view=powershell-7.3#example-9-create-a-detached-process-on-linux).

We might want to add MacOS to the matrix and thereby test all supported/OOTB operating systems and later maybe also use [service containers](https://learn.microsoft.com/en-us/azure/devops/pipelines/process/service-containers?view=azure-devops&tabs=yaml) to test SQL Server on non-Windows OSes.

### CODE REVIEW

File: build/azure-pipelines.yml [hunk 1]
```diff
@@ -351,137 +345,160 @@ stages:
...
+            workingDirectory: $(Agent.BuildDirectory)/app
+
+          # Start SQL Server LocalDB
+          - pwsh: SqlLocalDB start MSSQLLocalDB
+            displayName: Start SQL Server LocalDB
+            condition: and(succeeded(), eq(variables['CONNECTIONSTRINGS__UMBRACODBDSN_PROVIDERNAME'], 'Microsoft.Data.SqlClient'), eq(variables['Agent.OS'], 'Windows_NT'))
+
+          # Run application
+          - bash: |
+              nohup dotnet run --project UmbracoProject --configuration $(buildConfiguration) --no-build --no-launch-profile > $(Build.ArtifactStagingDirectory)/playwright.log 2>&1 &
```
Comment (ronaldbarendse) [inline build/azure-pipelines.yml:440]: Instead of redirecting the console output, we could also copy the application logs (`$(Agent.BuildDirectory)/app/UmbracoProject/umbraco/Logs`) into the artifact directory. The JSON logs contain a bit more information/detail, but might be harder to read. Any thoughts?

File: build/azure-pipelines.yml [hunk 2]
```diff
@@ -351,137 +345,160 @@ stages:
...
+              dotnet restore UmbracoProject
+              dotnet build UmbracoProject --configuration $(buildConfiguration) --no-restore
+              dotnet dev-certs https
+            displayName: Build application
+            workingDirectory: $(Agent.BuildDirectory)/app
+
+          # Start SQL Server LocalDB
+          - pwsh: SqlLocalDB start MSSQLLocalDB
+            displayName: Start SQL Server LocalDB
+            condition: and(succeeded(), eq(variables['CONNECTIONSTRINGS__UMBRACODBDSN_PROVIDERNAME'], 'Microsoft.Data.SqlClient'), eq(variables['Agent.OS'], 'Windows_NT'))
```
Comment (ronaldbarendse) [inline build/azure-pipelines.yml:436]: Once a service container is used for SQL Server, we can potentially remove the usage of this LocalDB instance, since they're using the same database engine under the hood... The install/setup using LocalDB is slightly different though, so that part might still be valuable to test.

**Review by ronaldbarendse — COMMENTED** [review]
Looks like the Linux test is about 2 minutes quicker to finish than before (because it doesn't have to build the Docker image) and the Windows test is about the same (not that surprising, since the setup is almost the same)...

Slightly surprising is that the SQLite and SQL Server tests on Windows take almost the same time, although SQL Server is known to be faster than SQLite. I guess the main reason is that we're not running the tests concurrently, so we don't experience the single concurrent writer limitation/bottleneck of SQLite (ignoring the fact that starting the LocalDB instance takes about 40 seconds).

**Comment by ronaldbarendse** [comment]
Some additional thoughts/things we can look into:
- Using service containers to run SQL Server (as already mentioned): https://learn.microsoft.com/en-us/azure/devops/pipelines/process/service-containers?view=azure-devops&tabs=yaml
- Separating the build steps into templates: https://learn.microsoft.com/en-us/azure/devops/pipelines/process/templates?view=azure-devops
- Making it possible to conditionally remove stages (e.g. documentation and release steps on non-release builds): https://learn.microsoft.com/en-us/azure/devops/pipelines/security/templates?view=azure-devops#conditional-insertion-of-stages-or-jobs
- Update the Integration Tests to use a similar setup 😄

File: build/azure-pipelines.yml [hunk 3]
```diff
@@ -351,137 +345,160 @@ stages:
...
           - task: NodeTool@0
             displayName: Use Node.js $(nodeVersion)
             inputs:
               versionSpec: $(nodeVersion)
+
+          - task: UseDotNet@2
+            displayName: Use .NET $(dotnetVersion)
+            inputs:
+              version: $(dotnetVersion)
+              includePreviewVersions: $(dotnetIncludePreviewVersions)
```
Comment (ronaldbarendse) [inline build/azure-pipelines.yml:397]: We can also remove the .NET SDK version from the build pipeline by creating a [`global.json` file](https://learn.microsoft.com/en-us/dotnet/core/tools/global-json) in the repository root containing:

```json
{
  "sdk": {
    "version": "7.0.100",
    "rollForward": "latestFeature"
  }
}
```

And then remove the `dotnetVersion` and `dotnetIncludePreviewVersions` variables and replace all occurrences of this task to:

```suggestion
          - task: UseDotNet@2
            displayName: Use .NET SDK from global.json
            inputs:
              useGlobalJson: true
```

This will also ensure everyone that's running from source code will need to have the correct .NET SDK version installed (a minimum of the specified version and depending on the `rollForward` value, the same major/minor).

**Comment by bergmania** [comment]
Sounds good, but i would suggest to keep testing linux on sqlserver as optional

**Comment by ronaldbarendse** [comment]
> Sounds good, but i would suggest to keep testing linux on sqlserver as optional

SQL Server is currently only tested on Windows and takes the same amount of time to complete as SQLite, so it doesn't add to the total time (as the tests are run in parallel).

We could do something similar as the integration tests though, were some tests run conditionally based on branch (only release branches) or manually requested (using a build parameter), so we can reduce the resource utilization. @bergmania Should I add a new `sqlServerAcceptanceTests` parameter to the build to prevent running 'Windows (SQL Server)', making that opt-in already?

**Comment by bergmania** [comment]
> Should I add a new sqlServerAcceptanceTests parameter to the build to prevent running 'Windows (SQL Server)', making that opt-in already?

Yeah that was what I had in mind :muscle:

File: build/azure-pipelines.yml (hunk 3 above)
Comment (ronaldbarendse) [reply build/azure-pipelines.yml]: I've merged PR https://github.com/umbraco/Umbraco-CMS/pull/14720 into this PR, so this is now fixed as well!

**Comment by ronaldbarendse** [comment]
I've updated this PR to contain the changes done in:
- https://github.com/umbraco/Umbraco-CMS/pull/14720 - removing the need to specify the .NET SDK version in the build variables
- https://github.com/umbraco/Umbraco-CMS/pull/15319 - to reduce the `build_output` artifact size from ~1300MB to ~260MB

Removing the Docker build and thereby also the image from the E2E artifacts saves ~500MB, so this PR reduces the total artifact size _of every build_ by a whopping 1.5GB!

Before this PR, the following artifacts would be published (note that the Windows E2E test had some flaky tests that resulted in additional trace files, causing the artifact size to be 28MB):
Before PR artifacts

And after this PR, the artifact sizes are greatly reduced and E2E tests using SQL Server are automatically done for release builds or if the `Run SQL Server Acceptance Tests` parameter is manually enabled:
After PR artifacts

As mentioned before, we should split the build steps into multiple files/templates, as the `azure-pipelines.yml` is quite big and contains some repeated steps. While doing so, we can also make the setup for nightly builds much easier, since that could just use the templates directly... And finally, the build triggers could be moved into the build file as well, since that is now overridden in Azure Pipelines with manually configured triggers using branch filters.

**Comment by Zeegaan** [comment]
Still need to Adress Bjarkes comment, but the rest looks fine 😁

File: global.json [hunk 4]
```diff
@@ -1,7 +1,6 @@
 {
   "sdk": {
-    "version": "8.0.0",
-    "rollForward": "latestFeature",
-    "allowPrerelease": false
+    "version": "8.0.100",
```
Comment (bergmania) [inline global.json:3]: Why is this necessary? We have seen issue if we do not use 8.0.0, even that the version officially is 8.0.100

Comment (ronaldbarendse) [reply global.json:3]: What issue specifically? If any, it's probably related to pre-release versions, as 8.0.0 is lower than e.g. 8.0.100-rc.2... But if we want to ensure you're using the final version of the .NET 8 SDK, we should really set this to 8.0.100.

**Comment by ronaldbarendse** [comment]
I've [manually started a build](https://dev.azure.com/umbraco/Umbraco%20Cms/_build/results?buildId=152754&view=results) with both SQL Server tests enabled for the Integration Tests and E2E/Acceptance tests, which resulted in all green checks:

Successful build

So if the changes are reviewed/approved, this PR can get merged (don't forget to merge up to `v13/dev` and `v14/dev`, so any PRs targeting those branches will have smaller and faster builds) 👍🏻

File: global.json (hunk 4 above)
Comment (bergmania) [reply global.json:3]: I just remember it as built issues..That was the reason for this: https://github.com/umbraco/Umbraco-CMS/pull/15303

When using 8.0.0 and rollForward it seems to work for everyone.
This is also what Rider's UI generates when selecting .NET 8

Comment (ronaldbarendse) [reply global.json:3]: Could the build issues be caused by only having a preview/RC version of .NET 8 SDK installed? Using version 8.0.0 + lastestFeature will satisfy prerelease versions like `8.0.100-rc.2.23502.2` (although you might also need to enable `allowPrerelease` in that case). Version 8.0.100 is the first stable .NET 8 SDK version that was released, so why not explicitly use that as the minimum version?

Currently, version 8.0.101 has been released that contains a security patch, so maybe we should require everyone to update to avoid being vulnerable (similar to how we upgrade transient NuGet dependencies)?

**Review by iOvergaard — APPROVED** [review]
